### PR TITLE
[SQL] Handle modulo operations with negative values

### PR DESCRIPTION
**Pull request by abhizer**
Is this a user-visible change (yes/no): no

Should be merged after #1179 

Fixes: #1195

### CODE REVIEW

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/functions/FunctionsTest.java [hunk 1]
```diff
@@ -38,6 +38,40 @@ public void issue1180() {
         );
     }
 
+    // this is an edge case for negative integer modulo
+    // Rust's default modulo operator currently panics on
+    // INT::MIN / -1 instead of just returning 0
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/functions/FunctionsTest.java:43]: Rust panics for division, but this is modulo. What is the relationship?

**Review by mihaibudiu — CHANGES_REQUESTED** [review]

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/functions/FunctionsTest.java (hunk 1 above)
Comment (abhizer) [reply sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/functions/FunctionsTest.java:43]: `/` is a typo, it should be `%`

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/functions/FunctionsTest.java [hunk 2]
```diff
@@ -39,6 +39,40 @@ public void issue1192() {
         this.runtimeFail("select '9223372036854775808'::int64", "ParseIntError", this.getEmptyIOPair());
     }
 
+    // this is an edge case for negative integer modulo
+    // Rust's default modulo operator currently panics on
+    // INT::MIN % -1 instead of just returning 0
+    @Test
+    public void issue1187modulo() {
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/functions/FunctionsTest.java:46]: Let's add 4 more tests: 8 % 3, 8 % -3, -8 % 3 and -8 % -3.

File: docs/sql/integer.md [hunk 3]
```diff
@@ -10,6 +10,9 @@ The legal operations are `+` (plus, unary and binary), `-` (minus,
 unary and binary), `*` (multiplication), `/` (division), `%`
 (modulus).
 
+The modulo operator treats modulo by -1 (`x % -1`) as a special
```
Comment (mihaibudiu) [inline docs/sql/integer.md:13]: This is not really a special case. What should be here is a definition of what happens for negative numbers. Just showing the result for 8 % 3, 8 % -3, -8 % 3 and -8 % -3 should be enough.

**Review by mihaibudiu — CHANGES_REQUESTED** [review]

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java [hunk 4]
```diff
@@ -418,6 +418,19 @@ public VisitDecision preorder(DBSPBinaryExpression expression) {
                     result = right;
                 }
             }
+        } else if (expression.operation.equals(DBSPOpcode.MOD)) {
+            if (right.is(DBSPLiteral.class)) {
+                DBSPLiteral rightLit = right.to(DBSPLiteral.class);
+                IsNumericType iRightType = rightType.to(IsNumericType.class);
+                if (iRightType.isOne(rightLit)) {
+                    result = iRightType.getZero();
+                } else if (iRightType.isZero(rightLit)) {
```
Comment (abhizer) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java:427]: Unsure as to how to check for -1. First intuition was to add `getNegativeOne` method to the `IsNumericType` interface, but it is also implemented for `usize` types.

Comment (mihaibudiu) [reply sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java:427]: One way to do it is to check separately for each type.

**Review by mihaibudiu — APPROVED** [review]

**Comment by mihaibudiu** [comment]
Please resolve the conflicts and we'll merge this.
Hopefully it won't be too bad.
You don't need to implement the warning for -1 for division or modulo in this PR.

**Pull request merged**